File: react-native/ios/BluetoothControlCommandService.swift
//
//  BluetoothControlCommandService.swift
//  ThetaBleClientReactNative
//
//  Created on 2024/12/20.
//

import THETABleClient

let KEY_DEVICE = "device"
let KEY_MAC_ADDRESS = "macAddress"
let KEY_PAIRING = "pairing"

enum BluetoothControlCommandService {
    static func scanPeripheralDevice(id: Int,
                                     timeout: Int32,
                                     resolve: @escaping ([[String: Any?]]) -> Void,
                                     reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.bluetoothControlCommand else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        Task {
            do {
                let value = try await service.scanPeripheralDevice(timeout: timeout)
                resolve(fromTheta(peripheralDeviceList: value))
            } catch {
                reject(ERROR_TITLE, error.localizedDescription, error)
            }
        }
    }

    static func scanPeripheralDeviceStart(id: Int,
                                          timeout: Int32,
                                          sendEvent: @escaping ([String: Any]) -> Void,
                                          resolve: @escaping RCTPromiseResolveBlock,
                                          reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.bluetoothControlCommand else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        Task {
            do {
                class Callback: BluetoothControlCommandScanCallback {
                    let deviceId: Int
                    let sendEvent: ([String: Any]) -> Void
                    init(deviceId: Int, sendEvent: @escaping ([String: Any]) -> Void) {
                        self.deviceId = deviceId
                        self.sendEvent = sendEvent
                    }

                    func onCompleted(peripheralDeviceList: [PeripheralDevice]) {
                        sendEvent(
                            toNotify(
                                deviceId: deviceId,
                                characteristic: BleCharacteristic.scanBluetoothPeripheralDevice,
                                params: fromTheta(peripheralDeviceList: peripheralDeviceList),
                                error: nil
                            )
                        )
                    }

                    func onNotify(peripheralDevice: PeripheralDevice) {
                        sendEvent(
                            toNotify(
                                deviceId: deviceId,
                                characteristic: BleCharacteristic.notificationScannedBluetoothPeripheralDevice,
                                params: fromTheta(peripheralDevice: peripheralDevice),
                                error: nil
                            )
                        )
                    }
                }
                try await service.scanPeripheralDeviceStart(timeout: timeout, callback: Callback(deviceId: id, sendEvent: sendEvent))
                resolve(nil)
            } catch {
                reject(ERROR_TITLE, error.localizedDescription, error)
            }
        }
    }

    static func scanPeripheralDeviceStop(id: Int,
                                         resolve: @escaping RCTPromiseResolveBlock,
                                         reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.bluetoothControlCommand else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        Task {
            do {
                try await service.scanPeripheralDeviceStop()
                resolve(nil)
            } catch {
                reject(ERROR_TITLE, error.localizedDescription, error)
            }
        }
    }

    static func connectPeripheralDevice(id: Int,
                                        macAddress: String,
                                        resolve: @escaping RCTPromiseResolveBlock,
                                        reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.bluetoothControlCommand else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        Task {
            do {
                try await service.connectPeripheralDevice(macAddress: macAddress)
                resolve(nil)
            } catch {
                reject(ERROR_TITLE, error.localizedDescription, error)
            }
        }
    }

    static func deletePeripheralDevice(id: Int,
                                       macAddress: String,
                                       resolve: @escaping RCTPromiseResolveBlock,
                                       reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.bluetoothControlCommand else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        Task {
            do {
                try await service.deletePeripheralDevice(macAddress: macAddress)
                resolve(nil)
            } catch {
                reject(ERROR_TITLE, error.localizedDescription, error)
            }
        }
    }
}

func fromTheta(peripheralDeviceList: [PeripheralDevice]) -> [[String: Any?]] {
    peripheralDeviceList.map { peripheralDevice in
        fromTheta(peripheralDevice: peripheralDevice)
    }
}

func fromTheta(peripheralDevice: PeripheralDevice) -> [String: Any?] {
    [
        KEY_DEVICE: peripheralDevice.device,
        KEY_MAC_ADDRESS: peripheralDevice.macAddress,
        KEY_PAIRING: peripheralDevice.pairing,
        PeripheralDeviceStatus.companion.keyName: peripheralDevice.status.name,
    ]
}

File: react-native/android/src/main/java/com/ricoh360/thetableclientreactnative/BluetoothControlCommandService.kt
package com.ricoh360.thetableclientreactnative

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.WritableArray
import com.facebook.react.bridge.WritableMap
import com.ricoh360.thetableclient.BleCharacteristic
import com.ricoh360.thetableclient.service.BluetoothControlCommand
import com.ricoh360.thetableclient.service.data.PeripheralDevice
import com.ricoh360.thetableclient.service.data.values.PeripheralDeviceStatus

const val KEY_DEVICE = "device"
const val KEY_MAC_ADDRESS = "macAddress"
const val KEY_PAIRING = "pairing"

object BluetoothControlCommandService {
  suspend fun scanPeripheralDevice(id: Int, timeout: Int, promise: Promise) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      device ?: let {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.bluetoothControlCommand
      service ?: let {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }
      val result = service.scanPeripheralDevice(timeout)
      promise.resolve(fromTheta(result))
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }

  suspend fun scanPeripheralDeviceStart(
    id: Int,
    timeout: Int,
    promise: Promise,
    sendNotifyEvent: (WritableMap) -> Unit,
  ) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      device ?: let {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.bluetoothControlCommand
      service ?: let {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }
      service.scanPeripheralDeviceStart(timeout, object: BluetoothControlCommand.ScanCallback {
        override fun onCompleted(peripheralDeviceList: List<PeripheralDevice>) {
          super.onCompleted(peripheralDeviceList)
          sendNotifyEvent(
            toNotify(
              id,
              BleCharacteristic.SCAN_BLUETOOTH_PERIPHERAL_DEVICE,
              fromTheta(peripheralDeviceList),
              null,
            ),
          )
        }

        override fun onNotify(peripheralDevice: PeripheralDevice) {
          super.onNotify(peripheralDevice)
          sendNotifyEvent(
            toNotify(
              id,
              BleCharacteristic.NOTIFICATION_SCANNED_BLUETOOTH_PERIPHERAL_DEVICE,
              fromTheta(peripheralDevice),
              null,
            ),
          )
        }
      })
      println("before resolve")
      promise.resolve(null)
      println("after resolve")
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }

  suspend fun scanPeripheralDeviceStop(id: Int, promise: Promise) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      device ?: let {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.bluetoothControlCommand
      service ?: let {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }
      service.scanPeripheralDeviceStop()
      promise.resolve(null)
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }

  suspend fun connectPeripheralDevice(id: Int, macAddress: String, promise: Promise) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      device ?: let {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.bluetoothControlCommand
      service ?: let {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }
      service.connectPeripheralDevice(macAddress)
      promise.resolve(null)
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }

  suspend fun deletePeripheralDevice(id: Int, macAddress: String, promise: Promise) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      device ?: let {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.bluetoothControlCommand
      service ?: let {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }
      service.deletePeripheralDevice(macAddress)
      promise.resolve(null)
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }
}

fun fromTheta(peripheralDevice: PeripheralDevice): WritableMap {
  val result = Arguments.createMap()
  result.putString(KEY_DEVICE, peripheralDevice.device)
  result.putString(KEY_MAC_ADDRESS, peripheralDevice.macAddress)
  result.putBoolean(KEY_PAIRING, peripheralDevice.pairing)
  result.putString(PeripheralDeviceStatus.keyName, peripheralDevice.status.name)
  return result
}

fun fromTheta(peripheralDeviceList: List<PeripheralDevice>): WritableArray {
  val result = Arguments.createArray()
  peripheralDeviceList.forEach {
    result.pushMap(fromTheta(it))
  }
  return result
}
